dask/highlevelgraph.py

```diff
@@ -1,10 +1,103 @@
 from collections.abc import Mapping
+from typing import Set, Dict
+import copy
 
 import tlz as toolz
 
 from .utils import ignoring
 from .base import is_dask_collection
-from .core import reverse_dict
+from .core import reverse_dict, get_dependencies, flatten
+
+
+class Layer(Mapping):
+    """ High level graph layer
+
+    This abstract class establish a protocol for high level graph layers.
+    """
+
+    def cull(self, keys: Set) -> "Layer":
+        """ Return a new Layer with only the tasks required to calculate `keys`.
+
+        In other words, remove unnecessary tasks from the layer.
+
+        Examples
+        --------
+        >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})
+        >>> dsk = d.cull({'out'})
+        >>> dsk  # doctest: +SKIP
+        {'x': 1, 'out': (add, 'x', 10)}
+
+        Returns
+        -------
+        layer: Layer
+            Culled layer
+        """
+
+        seen = set()
+        out = {}
+        work = set(flatten(keys))
+        while work:
+            new_work = set()
+            for k in work:
+                out[k] = self[k]
+                for d in get_dependencies(self, k):
+                    if d not in seen:
+                        seen.add(d)
+                        new_work.add(d)
+            work = new_work
+
+        return BasicLayer(out)
+
+    def get_external_dependencies(self, known_keys: Set) -> Set:
```

Comment:
I think that it would be interesting / educational to implement these methods for Blockwise.  I think that engaging some complex graph layer type will help us constrain development here.

Reply:
Done :)